fixes #14638 - Refactor Subnet into STI to allow different subnet types

timogoebel opened the pull request:
This basically makes subnet a STI and introduces a `Subnet::Ipv4` child class.
What I want to be able to do is call `Subnet.new` and get a Subnet::Ipv4-Object if type is not passed.

This commit also adds a setter for cidr, changes the subnet factory to subnet_ipv4 and changes the ip validation code.

This is based on @domcleal 's ipv6 branch: develop...domcleal:2315-ipv6 . Kudos to him.

timogoebel committed:
fixes #14638 - Refactor Subnet into STI to allow different subnet types

ohadlevy commented on app/models/subnet.rb:
```diff
@@ -102,22 +115,41 @@ def <=>(other)
   def self.subnet_for(ip)
-    Subnet.all.each {|s| return s if s.contains? IPAddr.new(ip)}
+    ip = IPAddr.new(ip)
+    Subnet.all.each {|s| return s if s.family == ip.family && s.contains?(ip)}
```
nitpick, can change to use detect instead.

ohadlevy commented on app/models/subnet/ipv4.rb:
```diff
@@ -0,0 +1,42 @@
+
+class Subnet::Ipv4 < Subnet
+  def self.default_sti_class
+    'Subnet::Ipv4'
```
is this really required? I would assume its only needed on the top level parent class?

timogoebel replied:
I agree, that's a leftover from the IPv6 branch and can safely be deleted from the IPv4 class.

The code currently has the problem, that when you create a new instance of Subnet::Ipv6 without a type provided, it get's casted to Subnet::Ipv4 because `self.default_sti_class` is inherited to Subnet::Ipv6 as "Subnet::Ipv4". Let me see, what can be done against this...

dLobatog commented on app/models/concerns/foreman/sti.rb:
```diff
@@ -12,8 +12,9 @@ module STI
       def new_with_cast(*attributes, &block)
-        if (h = attributes.first).is_a?(Hash) && (type = h.with_indifferent_access.delete(:type)) && type.length > 0
-          if (klass = type.constantize) != self
+        default_type = self.respond_to?(:default_sti_class) && self.default_sti_class
```
This (`default_sti_class`) is not tested, please add a test for this

timogoebel replied:
Do you mean add tests for this concern at all? I added tests for the behaviour in `test/unit/subnet.rb`.

dLobatog commented on app/models/subnet.rb:
```diff
@@ -1,17 +1,37 @@
+    :subnet_parameters_attributes, :cidr
+
+  def self.inherited(child)
+    child.instance_eval do
```
What is this needed for?

timogoebel replied:
This sets the modal name for the STI subclasses to "Subnet". It would be e.g. 'subnet_ipv4' with out this and break paths, forms, etc.
Does an inline comment for this makes sense?

dLobatog commented on app/models/subnet.rb:
```diff
@@ -238,43 +269,51 @@ def self.parse_dhcp_options(options)
   end
 
   def cleanup_ip(address)
     address.gsub!(/\.\.+/, ".")
```
Is this relevant for IPv6? I don't think they can be represented with dots so maybe this method should go to the subclass

timogoebel replied:
Would you then completely remove the cleanup for ipv6, i.e. move `cleanup_addresses` to the subclass?

ekohl replied:
IPv6 can be cleaned since you can write `:0001:` which is usually shortened to `:1:`. Maybe this is the right place to do so as well?

timogoebel replied:
Normalization is already done in `Net::Validations.normalize_ip`. Ruby however, does not accept these addresses you named.

```
IPAddr.new(':1:')
IPAddr::InvalidAddressError: invalid address
```

ekohl replied:
It was a simple part of it. I assume normalize_ip should do then and there is no cleanup needed. I also think it's sufficient to show an error on invalid input and not try to fix input, especially for IPv6.

timogoebel replied:
I fully agree. 👍

dLobatog commented on db/migrate/20160414063050_add_sti_to_subnets.rb:
```diff
@@ -0,0 +1,11 @@
+class AddStiToSubnets < ActiveRecord::Migration
+  def self.up
+    add_column :subnets, :type, :string, :default => 'Subnet::Ipv4'
+    execute "UPDATE subnets set type='Subnet::Ipv4'"
```
Why not `.type = Subnet::Ipv4`?

timogoebel replied:
This breaks migrations:

```
-- execute("UPDATE subnets.type = 'Subnet::Ipv4'")
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::Error: ERROR:  syntax error at or near "="
LINE 1: UPDATE subnets.type = 'Subnet::Ipv4'
```

dLobatog commented on app/models/subnet.rb:
```diff
@@ -23,34 +43,27 @@ class Subnet < ActiveRecord::Base
-                      :allow_nil => true,
-                      :format => {:with => Net::Validations::IP_REGEXP},
-                      :length => { :maximum => 15, :message => N_("is too long (maximum is 15 characters)") }
-  validates :mask,    :format => {:with => Net::Validations::MASK_REGEXP}
```
Where did these 3 validations go?

timogoebel replied:
They're in `ensure_ip_addrs_valid`.

dLobatog commented on app/models/subnet/ipv4.rb:
```diff
@@ -0,0 +1,38 @@
+  has_many :interfaces, :class_name => 'Nic::Base', :foreign_key => :subnet_id
+  has_many :primary_interfaces, -> { where(:primary => true) }, :class_name => 'Nic::Base', :foreign_key => :subnet_id
+  has_many :hosts, :through => :interfaces
+  has_many :primary_hosts, :through => :primary_interfaces, :source => :host
```
Why moving these last 2 has_many? I can see why the other 2 need to be moved, because `:foreign_key` can also be `:subnet6_id`, but these two don't have that limitation

timogoebel replied:
If I don't move them (read: leave them in the Subnet class), rails complains.
Is there anything I can do about this?

Apparently, you have to define the "has_many" before the "has_many through"....

```
ActiveRecord::HasManyThroughAssociationNotFoundError: Could not find the association :interfaces in model Subnet
```

dLobatog replied:
Interesting, I suppose it has something to do with class loading order. Just leave a comment so we know it and we're good!

dLobatog commented on app/models/subnet.rb:
```diff
@@ -1,17 +1,37 @@
+    'Subnet::Ipv4' if self.name == 'Subnet'
+  end
+
+  IP_FIELDS = [:network, :mask, :gateway, :dns_primary, :dns_secondary, :from, :to]
```
It looks like it would make sense to have an `Ip` class that hold all the validations, changes, etc...

timogoebel replied:
Probably yes, let's open a Redmine for this. If I or somebody else finds the time we can improve this. The code in this PR might not be perfect, but i think it's an improvement. I definitely don't have the time to refactor this even more...

timogoebel commented:
@dLobatog : Thank you for your comments. I have changed the default casting a little bit, to not violate the KISS-principle (see `new` in app/models/subnets.rb) and addressed your comments. Splitting the PR has definitely payed off already. Divide and conquer. I have two questions back though. Please see the inline comments. Could you have another look at the code and tell me what needs to be done to get this merged. Thanks for baring with me here.

ares commented on db/migrate/20160414063050_add_sti_to_subnets.rb:
```diff
@@ -0,0 +1,11 @@
+class AddStiToSubnets < ActiveRecord::Migration
+  def self.up
+    add_column :subnets, :type, :string, :default => 'Subnet::Ipv4'
```
if you set type `:null => false` which I'd recommend anyway, I think you don't have to explicitly do the update query, it should set the default value

timogoebel replied:
@ares: Yes, indeed. That's better. Thanks!

ares commented on lib/net/validations.rb:
```diff
@@ -24,9 +31,16 @@ def valid_mac?(mac)
+    end
+
+    # validates an IPv4 address and raises an error
+    def validate_ip! ip
```
nit: argument should be wrapped in `()`

ares commented on lib/net/validations.rb:
```diff
@@ -9,6 +12,10 @@ module Validations
     end
 
+    class << self
+      include Net::Validations
```
I'm not sure why this is needed but I'm not sure it's a good idea to make **all** instance methods available as module functions. Why not adding specific ones by `module_function` like existing `valid_mac?`?

timogoebel replied:
I believe, this is not necessary at all. I changed the methods to just use "def self.abc". I think, that's the best approach.

dLobatog commented on app/models/subnet.rb:
```diff
@@ -102,22 +120,40 @@ def <=>(other)
   def cidr
+    return if mask.nil?
     IPAddr.new(mask).to_i.to_s(2).count("1")
+  rescue ArgumentError
```
I guess if you rescue this is because we call `cidr(something)` and you want to discard that. Why/when does this happen?

timogoebel replied:
@dLobatog : This is necessary when mask holds an invalid value, e.g. the user enters "green" as mask. Same below.

dLobatog replied:
But that's another kind of exception isn't it?

``` ruby
IPAddr.new('green')
IPAddr::InvalidAddressError: invalid address
```

timogoebel replied:
@dLobatog : That's correct. I had that in there before, can't remember why I had to change that to ArgumentError. I changed the rescue to "IPAddr::InvalidAddressError". We'll see if and when that'll be a problem.

timogoebel replied:
@dLobatog : On Ruby 1.9 `IPAddr::InvalidAddressError` is not defined. IPAddr throws an ArgumentError there. How should we proceed?

timogoebel replied:
@dLobatog : I added a compatibility layer for ruby 1.9. Please see the code.

dLobatog commented on app/models/subnet.rb:
```diff
@@ -102,22 +120,40 @@ def <=>(other)
+  def cidr=(cidr)
+    return if cidr.nil?
+    self[:mask] = IPAddr.new(in_mask, family).mask(cidr).to_s
+  rescue ArgumentError
```
What ArgumentError?

dLobatog commented on test/unit/subnet/ipv4_test.rb:
```diff
@@ -127,6 +133,13 @@ def create_a_domain_with_the_subnet
+    s=subnets(:one)
+    s.from = "2.3.4.abc"
+    s.to   = "2.3.4.17"
+    assert !s.save
```
nitpick, `refute s.valid?` (saves 1 call to db)

dLobatog replied:
Or even `should_not allow_value('2.3.4.abc').for(:from)` , shoulda-matchers syntax.

dLobatog commented on test/unit/subnet.rb:
```diff
@@ -0,0 +1,19 @@
+require 'test_helper'
+
+class SubnetTest < ActiveSupport::TestCase
+  test 'should be casted to Subnet::Ipv4 if no type is set' do
```
Nitpick on the test description, "cast" past participle -> "cast"

dLobatog commented:
I just left a few minor comments, nitpicks mostly. Other than that, I think it should be ready to roll. Thanks @timogoebel !

ekohl commented on lib/net/validations.rb:
```diff
@@ -22,43 +25,45 @@ def valid_mac?(mac)
     end
 
-    def validate_mask(mask)
+    def self.validate_mask(mask)
```
Why do you have `validate_ip!` but not `validate_mask!` (and the same for mac + hostname)?

timogoebel replied:
@ekohl: To be honest, I just copied this from Dominic's branch without giving it much thought. The diff was already way too big for my taste and I tried to ignore everything not really necessary. I have cleaned the interface up though and have added some tests. Mind to take brief look?

timogoebel commented:
@dLobatog : Fixed the nitpicks, left an inline commend, made the Net::Valdidations-Interface clean and added some missing tests for the validations.

dLobatog commented:
@timogoebel Awesome, :+1:  from me if tests pass

timogoebel commented:
@dLobatog : Tests passed. :-)

dLobatog commented:
Merged as 05848084edcd1a4edfbddab8ac2bd46c66c77b4c, thanks @timogoebel!
